### PR TITLE
Fix krb5 library SO name in the gss api shim

### LINKED ISSUE #59518: Kerberos authentication not work on .NET 6 RC1 with official docker images
### Description
Kerberos authentication doesn't work after update to RC1 from .NET 5.0:
```**Cannot load library libgssapi_krb5.so**
Error: libgssapi_krb5.so: cannot open shared object file: No such file or directory
Unhandled exception. System.ComponentModel.Win32Exception (0x80090020): The type initializer for 'NetSecurityNative' threw an exception.
   at System.Net.Security.NegotiateStreamPal.AcquireCredentialsHandle(String package, Boolean isServer, NetworkCredential credential)
   at System.Net.NTAuthentication.Initialize(Boolean isServer, String package, NetworkCredential credential, String spn, ContextFlagsPal requestedContextFlags, ChannelBinding channelBinding)
   at System.Net.Http.AuthenticationHelper.SendWithNtAuthAsync(HttpRequestMessage request, Uri authUri, Boolean async, ICredentials credentials, Boolean isProxyAuth, HttpConnection connection, HttpConnectionPool connectionPool, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendUsingHttp11Async(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.DetermineVersionAndSendAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendAndProcessAltSvcAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.AuthenticationHelper.SendWithAuthAsync(HttpRequestMessage request, Uri authUri, Boolean async, ICredentials credentials, Boolean preAuthenticate, Boolean isProxyAuth, Boolean doRequestAuth, HttpConnectionPool pool, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageHandlerStage.Send(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageHandlerStage.Send(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.SocketsHttpHandler.Send(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageInvoker.Send(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.Send(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.Send(HttpRequestMessage request)
```
Reproduced on mcr.microsoft.com/dotnet/aspnet:6.0 and mcr.microsoft.com/dotnet/aspnet:6.0-focal.

### Configuration
```
dotnet --info

Host (useful for support):
  Version: 6.0.0-rc.1.21451.13
  Commit:  d7619cd4b1

.NET SDKs installed:
  No SDKs were found.

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.0-rc.1.21452.15 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

### Other information
Workaround:
```
ln /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so
```

Caused by  Shim gss api on Linux to delay loading libgssapi_krb5.so #55037

### PR DESCRIPTION
**Pull request by janvorli**
The library name used in the shim is a name of the build time library which
is usually installed only for development purposes. We should use
libgssapi_krb5.so.2 which is the underlying runtime library.

Close #59518

### CODE REVIEW

**Review by karelz — APPROVED** [review]

**Comment by VSadov** [comment]
I assumed it is the purpose of libgssapi_krb5.so - to link to the actual underlying library, assuming there are different implementations of the api with different names, i.e. MIT vs. Heimdall.

Do they both just install “so.2”?

File: src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 1]
```diff
@@ -123,7 +123,7 @@ static void* volatile s_gssLib = NULL;
 #define GSS_C_NT_HOSTBASED_SERVICE          (*GSS_C_NT_HOSTBASED_SERVICE_ptr)
 #define gss_mech_krb5                       (*gss_mech_krb5_ptr)
 
-#define gss_lib_name "libgssapi_krb5.so"
+#define gss_lib_name "libgssapi_krb5.so.2"
```
Comment (am11) [inline src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:126]: Can we align it with how it's done for numa/ssl?

```
try dlopen libgssapi_krb5.so.2
then try dlopen libgssapi_krb5.so.2.2
then try dlopen libgssapi_krb5.so
```

Comment (janvorli) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:126]: For SSL, it is different as there are several incompatible versions of the library we support. We always use a versioned one though. But adding fallback to libgssapi_krb5.so sounds probably fine, even though I am not aware of a distro where it would be named that way and usually a change in the major version of the library means a compatibility break. So loading just the .so if present might result in hard to debug crashes in case e.g. libgssapi_krb5.so.3 was released and had some subtle incompatible changes.

**Comment by janvorli** [comment]
> I assumed it is the purpose of libgssapi_krb5.so - to link to the actual underlying library

No, in most cases, the .so without version is a dev time library. The linker then stores the actual library SONAME which includes the version to the target binary (when linking the library at build time).

File: src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c (hunk 1 above)
Comment (janvorli) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:126]: @am11, it actually seems that for the NUMA case, attempting to use the versionless .so is also something that we might not want to do for the same reason I've mentioned. And do you remember why you have also added the probe for libnuma.so.1.0.0? Have we found systems where the libnuma.so.1 was not present? The comment on the PR that added the libnuma probing is actually not right, there is also libnuma.so on Fedora, the package is numactl-devel.

**Comment by janvorli** [comment]
The only thing I am not sure about is FreeBSD. I have found I have two of the libraries installed there:
```
/usr/lib/libgssapi_krb5.so
/usr/lib/libgssapi_krb5.so.10
/usr/local/lib/libgssapi_krb5.so
/usr/local/lib/libgssapi_krb5.so.2
/usr/local/lib/libgssapi_krb5.so.2.2
```
Might be just a result of my past experiments. I wonder if the libgssapi_krb5.so.10 is the one to use on FreeBSD and so we should add probing for it.
@josteink, @Thefrank I can see you are heavily working on FreeBSD .NET stuff, what is the truth here?

**Comment by VSadov** [comment]
It seems like in general the most consistent way would be to mimic the static linker behavior - resolve the link at build time and use the actual versioned .so - not because it is strictly better, but just to match the common practice.

However, since we have just one case like this, if we are sure it is always so.2, then current fix is good enough.

**Comment by wfurt** [comment]
The `libgssapi_krb5.so.10`  is part of base OS and comes from Heimdal (same as macOS). The `libgssapi_krb5.so.2` comes from `krb5` package and it is the MIT implementation. 
Conceptually, dependency on MIT _could_ work but I would think using base OS would be preferable. 

Seems like the packaging strategy is different on Linux. In bot cases for Heimdal and MIT the `.so` is part of core package not specific to debug. If we load the `.so` than user can set LD_LIBRARY_PATH to choose implementation. (but I'm not sure how common that would be)

I'm wondering if we should scope this change to Linux only where the claim about development is probably true.

**Comment by wfurt** [comment]
BTW since this is regression, I would be ok if we proceed with this for 6.0.

**Comment by Thefrank** [comment]
TL;DR: The ports version (aka MIT version, the one found in `/usr/local/lib`) is what would be preferred. 

Long version:
Generally: items that ship with the OS are for the OS and items found in ports should use other items found in ports.

A specific example of this is that FreeBSD ships with a full LLVM suite. This is great until you want to use an `lldb` plugin only to find out plugins are not supported. The ports version of the very same LLVM suite includes this support. [As a side rant: this makes proper compiler detection a pain a times]

Also, I think we currently list `krb5` as an optional package dependency for net5.

**Comment by janvorli** [comment]
@Thefrank thank you a lot for your quick response! Then the current change is good for FreeBSD too.

**Review by VSadov — APPROVED** [review]

**Comment by Thefrank** [comment]
@janvorli I think this PR is getting bit by the "license/cla missing" bug

**Comment by janvorli** [comment]
@dotnet/dnceng why am I seeing the license/cla "waiting for status to be reported"? What is the way to fix it? Closing and reopening the PR?

**Comment by alexperovich** [comment]
A close and re-open will probably fix it. This is the cla bot that the opensource team manages. Re-opening the PR would re-trigger the web hooks that make the cla thing go. I haven't heard of any current issues with the bot or github.
